src/libcore/slice/mod.rs

```diff
@@ -3215,7 +3223,7 @@ macro_rules! iterator {
             fn next(&mut self) -> Option<$elem> {
                 // could be implemented with slices, but this avoids bounds checks
                 unsafe {
-                    assume(!self.ptr.is_null());
+                    assume(!self.ptr.as_ptr().is_null());
```

Comment:
This is now implied by the `NonNull` type, right? Should we remove this `assume`?

Reply:
My understanding is it's an optimization when it comes to code generation, not layout optimization. But maybe rustc is smart enough about that? I'm not that familiar with how `NonNull` works.

Reply:
[Looks like](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=f4417464a215e39c5c5ebccf85d6c733) we could remove it. Is that example good enough evidence? Or should we reach out to someone who knows better?

Reply:
The "this pointer is non-null" information added by rustc to LLVM IR automatically is limited to particular positions (function parameters, function return values, loads from memory) which can get eliminated by optimizations (such as inlining and SROA) before they have served their purpose. In contrast, assumes stay around basically for the entire optimization pipeline. So I am not confident that we can remove this assume without regressions. You're welcome to try to evaluate that on real world code (the example linked above is far too isolated, I would recommend history spelunking to find the benchmarks that motivated adding the assume in the first place) but I would recommend doing that in a separate PR.